### PR TITLE
Implement RotatingServerAdvice enhancements

### LINKED ISSUE #3027: Enhancements to RotatingServerAdvice
## Enhancement
Specific issues  that depends on this are that the OOTB Sftp Source app uses this class if configured for synchronizing remote files, but uses a parallel implementation for MessageSources that simply list the contents of the remote directory. Additionally, enhancements to the source require exposing the currentDirectory in all multi-source use cases. See https://github.com/spring-cloud-stream-app-starters/sftp/issues/60

1) `RotatingServerAdvice` delegates to `StandardRotationPolicy` which assumes that the MessageSource is one of two types, and downcasts the MessageSource to the appropriate type to perform some specific actions.  This should be decoupled from the generic server rotation logic to allow use with other type of message sources.  
2) Expose the `currentDirectory()`of RotationPolicy so subclasses of `RotatingServerAdvice` can access it.

### PR DESCRIPTION
**Pull request by dturanski**
Resolves #3027

Create `AbstractStandardRotationPolicy` subclasses implement `onRotation(MessageSource source)`.

Pull dependent inner classes from RotatingServerAdvice to break circular dependencies.

### CODE REVIEW

File: spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotationPolicy.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	void afterReceive(boolean messageReceived, MessageSource<?> source);
+
+
+	/**
+	 *
+	 * @return the current {@link KeyDirectory}
+	 */
+	KeyDirectory getCurrent();
+
+}
```
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotationPolicy.java]: Missing newline.

File: spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotationPolicy.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	 * Invoked after the message source receive() method.
+	 * @param messageReceived true if a message was received.
+	 * @param source the message source.
+	 */
+	void afterReceive(boolean messageReceived, MessageSource<?> source);
+
+
+	/**
+	 *
+	 * @return the current {@link KeyDirectory}
```
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotationPolicy.java]: I don't think this will pass checkstyle. Description needed.

**Review by garyrussell — CHANGES_REQUESTED** [review]

**Comment by garyrussell** [comment]
@dturanski Please rebase and re-push; I have fixed the .travis.yml for this branch.

**Comment by garyrussell** [comment]
Merged as 03b6316d30d1c87c92dbfa00bb54ad0b3ceb2427 with minor formatting fixes.

File: spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotatingServerAdvice.java [hunk 3]
```diff
@@ -88,167 +84,28 @@ public boolean beforeReceive(MessageSource<?> source) {
 		return result;
 	}
 
-	/**
-	 * Implementations can reconfigure the message source before and/or after
-	 * a poll.
-	 */
-	public interface RotationPolicy {
```
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotatingServerAdvice.java]: We have to revert this change. It is fully blown breaking change in the point version.
I'm OK with a decoupling in the current `5.2` (preparing some polishing PR), but I'm against any breaking changes for public API in the point release.

Does it make sense, @garyrussell , @dturanski ?

Thanks

Comment (dturanski) [reply spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotatingServerAdvice.java]: Technically you're right. But I think it would be ok if we default it. Then it won't break anything. I'd be surprised if anyone else has actually implemented this interface.

Comment (artembilan) [reply spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotatingServerAdvice.java]: We just can't guess from here. This interface is `public` in this place, therefore moving it somewhere else is a breaking change.
The first rule: don't break any public API in the point release!

Comment (dturanski) [reply spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotatingServerAdvice.java]: I see, because I moved it from the inner class.  Yes that is a breaking change.

Comment (artembilan) [reply spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotatingServerAdvice.java]: Good. So, @garyrussell , your turn now and then we will make a final decision!

**Comment by garyrussell** [comment]
Yes; I had a brain fart yesterday.

But, we do need to satisfy David's extensibility needs.

**Comment by artembilan** [comment]
> But, we do need to satisfy David's extensibility needs.

Right, so we leave it fixed in `5.2` and also see my polishing in the #3032

**Comment by artembilan** [comment]
Reverted: https://github.com/spring-projects/spring-integration/commit/f7032c7127cc790bc974f62cea33bbce0b1bb147
